internal/api/server/apirequest_log_test.go

```diff
@@ -0,0 +1,180 @@
+		{
+			email:    "server-admin",
+			endpoint: "users/list",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/users/list",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					UserEmail: "server-admin@test.edulinq.org",
+				},
+			},
+		},
+
+		// Valid permissions, APIRequestCourseUserContext request.
+		{
+			email:    "server-admin",
+			endpoint: "courses/users/list",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/courses/users/list",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					UserEmail: "server-admin@test.edulinq.org",
+					CourseID:  "course101",
+				},
+			},
+		},
+
+		// Valid permissions, APIRequestAssignmentContext request.
+		{
+			email:    "server-admin@test.edulinq.org",
+			endpoint: "courses/assignments/get",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/courses/assignments/get",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					UserEmail:    "server-admin@test.edulinq.org",
+					CourseID:     "course101",
+					AssignmentID: "hw0",
+				},
+			},
+		},
+
+		// Invalid permissions, APIRequestUserContext request.
+		{
+			email:           "course-student",
+			endpoint:        "users/list",
+			expectedLocator: "-041",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/users/list",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					UserEmail: "course-student@test.edulinq.org",
+				},
+				Locator: "-041",
+			},
+		},
+
+		// Invalid permissions, APIRequestCourseUserContext request.
+		{
+			email:           "course-student",
+			endpoint:        "courses/users/list",
+			expectedLocator: "-020",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/courses/users/list",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					UserEmail: "course-student@test.edulinq.org",
+					CourseID:  "course101",
+				},
+				Locator: "-020",
+			},
+		},
+
+		// Invalid permissions, APIRequestAssignmentContext request.
+		{
+			email:           "server-user",
+			endpoint:        "courses/assignments/get",
+			expectedLocator: "-040",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/courses/assignments/get",
+				Locator:  "-040",
+			},
+		},
+
+		// Valid permissions, invalid field request.
+		{
+			email:           "server-admin",
+			endpoint:        "courses/assignments/get",
+			fields:          map[string]any{"assignment-id": "zzz"},
+			expectedLocator: "-022",
+			expectedMetric: &stats.APIRequestMetric{
+				Endpoint: "/api/v03/courses/assignments/get",
+				CourseAssignmentEmailMetric: stats.CourseAssignmentEmailMetric{
+					CourseID:     "course101",
+					AssignmentID: "zzz",
+				},
+				Locator: "-022",
+			},
+		},
+	}
+
+	port, err := util.GetUnusedPort()
+	if err != nil {
+		test.Fatalf("Failed to get an unused port: '%v'.", err)
+	}
+
+	oldPort := config.WEB_HTTP_PORT.Get()
```

Comment:
Can you just use the standard test server like most of the API endpoint tests?
server_test.go needs very low level control,
but I don't think you need to do that here.